### DISCUSSION TITLE
How to efficiently change background styles for MAUI buttons?

### DISCUSSION THREAD

**Discussion by MartyIX** [microsoft/microsoft-ui-xaml#9858] [Q&A]
I have checked source code of MAUI and it does an [interesting thing](https://github.com/dotnet/maui/blob/9d9288a2f2972b998842049ea3905661c12b98cf/src/Core/src/Platform/Windows/ButtonExtensions.cs#L86-L104) when setting background for WinUI buttons:

```cs
public static void UpdateBackground(this Button platformButton, IButton button)
{
	var brush = button.Background?.ToPlatform();

	if (brush is null)
		platformButton.Resources.RemoveKeys(BackgroundResourceKeys);
	else
		platformButton.Resources.SetValueForAllKey(BackgroundResourceKeys, brush);

	platformButton.RefreshThemeResources(); // I'M INTERESTED IN THIS
}

static readonly string[] BackgroundResourceKeys =
{
	"ButtonBackground",
	"ButtonBackgroundPointerOver",
	"ButtonBackgroundPressed",
	"ButtonBackgroundDisabled",
};
```

And [RefreshThemeResources](https://github.com/dotnet/maui/blob/9d9288a2f2972b998842049ea3905661c12b98cf/src/Core/src/Platform/Windows/FrameworkElementExtensions.cs#L344-L356) is implemented as follows:

```cs
internal static void RefreshThemeResources(this FrameworkElement nativeView)
{
	var previous = nativeView.RequestedTheme;

	// Workaround for https://github.com/dotnet/maui/issues/7820
	nativeView.RequestedTheme = nativeView.ActualTheme switch
	{
		ElementTheme.Dark => ElementTheme.Light,
		_ => ElementTheme.Dark
	};

	nativeView.RequestedTheme = previous;
}
```

This implementation was introduced in https://github.com/dotnet/maui/issues/7820#issue-1263524481 which mentions

> There is a workaround and that is to toggle the theme, but a "real fix" is coming in https://dev.azure.com/microsoft/OS/_workitems/edit/39897009/

Obviously, I can't check that internal link so I'm interested if we can do now better given that 2 years passed since this was implemented.

Does anyone know how to improve this implementation? I ask because in my layout, 180 ms is taken by `RefreshThemeResources` calls.

**Comment by whiskhub** [comment]
This seems to be the related GitHub issue: https://github.com/microsoft/microsoft-ui-xaml/issues/5457
I don't think there has been any update, sadly.

**Reply by MartyIX** [reply]
Thank you! I read the discussion there. It's sad. Hopefully, things will improve over time.
